src/Castle.Core/DynamicProxy/CustomAttributeInfo.cs

```diff
@@ -0,0 +1,294 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Castle.DynamicProxy
+{
+	using System;
+	using System.Collections.Generic;
+	using System.Linq;
+	using System.Linq.Expressions;
+	using System.Reflection;
+	using System.Reflection.Emit;
+
+	public class CustomAttributeInfo : IEquatable<CustomAttributeInfo>
+	{
+		// Cached empty arrays to avoid unnecessary allocations
+		private static readonly PropertyInfo[] EmptyProperties = new PropertyInfo[0];
+		private static readonly FieldInfo[] EmptyFields = new FieldInfo[0];
+		private static readonly object[] EmptyValues = new object[0];
+
+		private static readonly IEqualityComparer<object> ValueComparer = new AttributeArgumentValueEqualityComparer();
+
+		private readonly CustomAttributeBuilder builder;
+		private readonly ConstructorInfo con;
+		private readonly object[] constructorArgs;
+		private readonly IDictionary<string, object> properties;
+		private readonly IDictionary<string, object> fields;
+
+		public CustomAttributeInfo(
+			ConstructorInfo con,
+			object[] constructorArgs,
+			PropertyInfo[] namedProperties,
+			object[] propertyValues,
+			FieldInfo[] namedFields,
+			object[] fieldValues)
+		{
+			// Will take care of validating the arguments
+			this.builder = new CustomAttributeBuilder(con, constructorArgs, namedProperties, propertyValues, namedFields, fieldValues);
+
+			this.con = con;
+			this.constructorArgs = constructorArgs;
+
+			this.properties = namedProperties
+				.Zip(propertyValues, (p, v) => new { p.Name, Value = v })
+				.ToDictionary(x => x.Name, x => x.Value);
+
+			this.fields = namedFields
+				.Zip(fieldValues, (f, v) => new { f.Name, Value = v })
+				.ToDictionary(x => x.Name, x => x.Value);
+		}
+
+		public CustomAttributeInfo(
+			ConstructorInfo con,
+			object[] constructorArgs,
+			PropertyInfo[] namedProperties,
+			object[] propertyValues)
+			: this(con, constructorArgs, namedProperties, propertyValues, EmptyFields, EmptyValues)
+		{			
+		}
+
+		public CustomAttributeInfo(
+			ConstructorInfo con,
+			object[] constructorArgs,
+			FieldInfo[] namedFields,
+			object[] fieldValues)
+			: this(con, constructorArgs, EmptyProperties, EmptyValues, namedFields, fieldValues)
+		{
+		}
+
+		public CustomAttributeInfo(
+			ConstructorInfo con,
+			object[] constructorArgs)
+			: this(con, constructorArgs, EmptyProperties, EmptyValues, EmptyFields, EmptyValues)
+		{
+		}
+
+		public static CustomAttributeInfo FromExpression(Expression<Func<Attribute>> expression)
```

Comment:
I'm sure this is perfectly correct, but it would be nice to see it exercised at least once in the tests.